Wrong error: return type references an anonymous lifetime

### Code

```Rust
pub async fn foo(_: Unknown<'_>) {}
```

### Current output

```Shell
error[E0412]: cannot find type `Unknown` in this scope
 --> src/main.rs:1:21
  |
1 | pub async fn foo(_: Unknown<'_>) {}
  |                     ^^^^^^^ not found in this scope
  |
help: there is an enum variant `std::os::unix::net::AncillaryError::Unknown`; try using the variant's enum
  |
1 | pub async fn foo(_: std::os::unix::net::AncillaryError) {}
  |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0581]: return type references an anonymous lifetime, which is not constrained by the fn input types
 --> src/main.rs:1:1
  |
1 | pub async fn foo(_: Unknown<'_>) {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: lifetimes appearing in an associated or opaque type are not considered constrained
  = note: consider introducing a named lifetime parameter

Some errors have detailed explanations: E0412, E0581.
For more information about an error, try `rustc --explain E0412`.
error: could not compile `stat_all` (bin "stat_all") due to 2 previous errors
```

### Desired output

```Shell
error[E0412]: cannot find type `Unknown` in this scope
 --> src/main.rs:1:21
  |
1 | pub async fn foo(_: Unknown<'_>) {}
  |                     ^^^^^^^ not found in this scope
  |
help: there is an enum variant `std::os::unix::net::AncillaryError::Unknown`; try using the variant's enum
  |
1 | pub async fn foo(_: std::os::unix::net::AncillaryError) {}
  |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

### Rationale and extra context

That error (the second one) of course disappears when `Unknown` gets into scope.
Note that this only happens with async, but not after manually desugaring it to `impl Future`.

### Other cases

```Rust

```

### Rust Version

```Shell
Latest nightly and 1.84
```

### Anything else?

_No response_

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs
@@ -2557,27 +2557,29 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         // reject function types that violate cmse ABI requirements
         cmse::validate_cmse_abi(self.tcx(), self.dcx(), hir_id, abi, bare_fn_ty);
 
-        // Find any late-bound regions declared in return type that do
-        // not appear in the arguments. These are not well-formed.
-        //
-        // Example:
-        //     for<'a> fn() -> &'a str <-- 'a is bad
-        //     for<'a> fn(&'a String) -> &'a str <-- 'a is ok
-        let inputs = bare_fn_ty.inputs();
-        let late_bound_in_args =
-            tcx.collect_constrained_late_bound_regions(inputs.map_bound(|i| i.to_owned()));
-        let output = bare_fn_ty.output();
-        let late_bound_in_ret = tcx.collect_referenced_late_bound_regions(output);
+        if !bare_fn_ty.references_error() {
+            // Find any late-bound regions declared in return type that do
+            // not appear in the arguments. These are not well-formed.
+            //
+            // Example:
+            //     for<'a> fn() -> &'a str <-- 'a is bad
+            //     for<'a> fn(&'a String) -> &'a str <-- 'a is ok
+            let inputs = bare_fn_ty.inputs();
+            let late_bound_in_args =
+                tcx.collect_constrained_late_bound_regions(inputs.map_bound(|i| i.to_owned()));
+            let output = bare_fn_ty.output();
+            let late_bound_in_ret = tcx.collect_referenced_late_bound_regions(output);
 
-        self.validate_late_bound_regions(late_bound_in_args, late_bound_in_ret, |br_name| {
-            struct_span_code_err!(
-                self.dcx(),
-                decl.output.span(),
-                E0581,
-                "return type references {}, which is not constrained by the fn input types",
-                br_name
-            )
-        });
+            self.validate_late_bound_regions(late_bound_in_args, late_bound_in_ret, |br_name| {
+                struct_span_code_err!(
+                    self.dcx(),
+                    decl.output.span(),
+                    E0581,
+                    "return type references {}, which is not constrained by the fn input types",
+                    br_name
+                )
+            });
+        }
 
         bare_fn_ty
     }

--- a/tests/ui/async-await/unconstrained-lifetimes.rs
+++ b/tests/ui/async-await/unconstrained-lifetimes.rs
@@ -1,0 +1,9 @@
+//@ edition: 2021
+
+// Make sure we don't complain about the implicit `-> impl Future` capturing an
+// unconstrained anonymous lifetime.
+
+async fn foo(_: Missing<'_>) {}
+//~^ ERROR cannot find type `Missing` in this scope
+
+fn main() {}

--- a/tests/ui/async-await/unconstrained-lifetimes.stderr
+++ b/tests/ui/async-await/unconstrained-lifetimes.stderr
@@ -1,0 +1,9 @@
+error[E0412]: cannot find type `Missing` in this scope
+  --> $DIR/unconstrained-lifetimes.rs:6:17
+   |
+LL | async fn foo(_: Missing<'_>) {}
+   |                 ^^^^^^^ not found in this scope
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0412`.

